ui/components/TracePage/PromptEditor.tsx

```diff
@@ -0,0 +1,236 @@
+      onClick={ev => {
+        if (!promptBoxRef.current?.contains(ev.target as Node)) {
+          closeEditor();
+        }
+      }}
+    >
+      <div
+        className="rounded-4 px-6 py-4 bg-white h-2/3 w-2/3 shadow cursor-default flex"
+        ref={promptBoxRef}
+      >
+        <div className="flex flex-col w-full">
+          <h3 className="text-lg font-semibold text-gray-700 mb-2">Prompt</h3>
+          <div
+            className="rounded-4 flex-grow border-slate-200 border px-3 py-2 whitespace-pre-wrap w-full overflow-auto"
+            contentEditable
+            onInput={ev => handleInput(ev)}
+            onKeyDown={ev => {
+              if (ev.key === "Enter") {
+                ev.preventDefault();
+                ev.stopPropagation();
+                document.execCommand("insertLineBreak");
+              }
+            }}
+            ref={promptContentRef}
+          />
+        </div>
+        <div className="flex flex-col w-48 pl-6">
+          <h3 className="text-lg font-semibold text-gray-700 mb-2">Agent</h3>
+          <label className="text-sm mb-1 text-gray-600">Model</label>
+          <Select onChange={ev => setSelectedAgent(ev.target.value)} value={selectedAgent}>
+            {agents.map(el => (
+              <option key={el} value={el}>
+                {el}
+              </option>
+            ))}
+          </Select>
+          <label className="text-sm mb-1 mt-3 text-gray-600">Multiline</label>
+          <Checkbox isChecked={multiline} onChange={ev => setMultiline(ev.target.checked)} />
+          <div className="flex-grow"></div>
+          <label className="text-sm mb-1 mt-3 text-gray-600">Keep prompt results</label>
+          <Checkbox isChecked={keepResults} onChange={ev => setKeepResults(ev.target.checked)} />
+          <div>
+            <Button
+              aria-label="prompt language model"
+              className="rounded-4 px-2 py-1 h-fit !shadow-none hover:bg-slate-200 mt-4 text-xs"
+              size="md"
+              variant="outline"
+              onClick={() => {
+                setLoading(true);
+                fetch(`${backendUrl}/agents/complete`, {
+                  method: "POST",
+                  headers: {
+                    "Content-Type": "application/json",
+                  },
+                  body: JSON.stringify({
+                    agent: selectedAgent,
+                    prompt: currentPrompt,
+                    multiline,
+                  }),
+                })
+                  .then(res => res.json())
+                  .then(data => {
+                    setPromptResult(data);
+                    setLoading(false);
+                  });
+              }}
+            >
+              {loading ? "Loading..." : "Prompt model"}
+            </Button>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
```

Comment:
This component is almost 150 lines long. Do you think there are some functions, hooks, or subcomponents that could be cleanly factored out?

Reply:
put the contenteditable out - that shouldn't be entangled but other stuff can be quite connected